source/extensions/network/dns_resolver/cares/dns_impl.cc

```diff
@@ -557,26 +556,29 @@ class CaresDnsResolverFactory : public DnsResolverFactory,
         new envoy::extensions::network::dns_resolver::cares::v3::CaresDnsResolverConfig()};
   }
 
-  DnsResolverSharedPtr createDnsResolver(Event::Dispatcher& dispatcher, Api::Api& api,
-                                         const envoy::config::core::v3::TypedExtensionConfig&
-                                             typed_dns_resolver_config) const override {
+  absl::StatusOr<DnsResolverSharedPtr>
+  createDnsResolver(Event::Dispatcher& dispatcher, Api::Api& api,
+                    const envoy::config::core::v3::TypedExtensionConfig& typed_dns_resolver_config)
+      const override {
     envoy::extensions::network::dns_resolver::cares::v3::CaresDnsResolverConfig cares;
     std::vector<Network::Address::InstanceConstSharedPtr> resolvers;
 
     ASSERT(dispatcher.isThreadSafe());
     // Only c-ares DNS factory will call into this function.
     // Directly unpack the typed config to a c-ares object.
-    THROW_IF_NOT_OK(Envoy::MessageUtil::unpackTo(typed_dns_resolver_config.typed_config(), cares));
+    RETURN_IF_NOT_OK(Envoy::MessageUtil::unpackTo(typed_dns_resolver_config.typed_config(), cares));
     if (!cares.resolvers().empty()) {
       const auto& resolver_addrs = cares.resolvers();
       resolvers.reserve(resolver_addrs.size());
       for (const auto& resolver_addr : resolver_addrs) {
         auto address_or_error = Network::Address::resolveProtoAddress(resolver_addr);
-        THROW_IF_STATUS_NOT_OK(address_or_error, throw);
+        RETURN_IF_STATUS_NOT_OK(address_or_error);
         resolvers.push_back(std::move(address_or_error.value()));
       }
     }
-    return std::make_shared<Network::DnsResolverImpl>(cares, dispatcher, resolvers,
+    auto csv_or_error = DnsResolverImpl::maybeBuildResolversCsv(resolvers);
```

Comment:
It looks to me there are some logic change here.   Previously, if resolvers is empty, the code still go ahead create DnsResolverImpl:  https://github.com/envoyproxy/envoy/blob/0d1ce088b8d87e03bb0a3dcbeececdd97ad716f3/source/extensions/network/dns_resolver/cares/dns_impl.cc#L58. 

Now we skipped it.

Reply:
I don't think so
we only fail if we have an error status.  in the case of empty  list, we still return absl::nullopt ("no resolvers") not absl::InvalidArgumentError  ("failure creating configured resolvers")

Reply:
I see. That static function DnsResolverImpl::maybeBuildResolversCsv() never returns Error.

But, it throw exceptions here:
https://github.com/envoyproxy/envoy/blob/9f494c0ba71f1a6d11864d8c4fc9afe298e4339a/source/extensions/network/dns_resolver/cares/dns_impl.cc#L67

Should we remove this throw?

Reply:
yeah I had set up to remove it and not yet removed it.  Happy to do so (fingers crossed on unit tests, else I'll pick back up monday)